src/nn/linear.rs

```diff
@@ -32,9 +32,11 @@ pub struct Linear<const I: usize, const O: usize> {
 }
 
 impl<const I: usize, const O: usize> CanUpdateWithGradients for Linear<I, O> {
-    fn update<G: GradientProvider>(&mut self, grads: &mut G) {
-        self.weight.update(grads);
-        self.bias.update(grads);
+    fn update<G: GradientProvider>(&mut self, grads: &mut G) -> Result<(), UnusedParamsError> {
+        let mut unused = Ok(());
+        unused.union(self.weight.update(grads).map_err(|l| l.prepend("weight")));
+        unused.union(self.bias.update(grads).map_err(|l| l.prepend("bias")));
+        unused
```

Comment:
This block feels like it could be improved somehow. the main core of it (self.weight.update & self.bias.update) is such a small portion of the code here.

Reply:
Does this mean that any user written module will have to manually map the error if they want to include it?

Also is there no qualifying name here? What happens when there are multipe layers that have the same parameter name?

Reply:
Yes as this PR is currently written, users would have to do the `map_err` thing and `union` if they wanted to properly compute the names of unused parameters.

How it works is the `l.prepend(...)` prepends the string to all parameters currently in `UnusedParamsError`. The union does a extend on the underlying vec. So in this case let's say both weight & bias were unused, it'd have 2 unused params. When this update call steps back up to the callee, it would also call .prepend() which would add more info to the parameters.

The example I was testing with was:
```rust
use dfdx::prelude::*;

fn main() {
    type Model = ((Repeated<Linear<5, 5>, 3>, ReLU), Linear<5, 4>);
    let mut model: Model = Default::default();

    // only use half of the model during forward
    let y = model.1.forward(Tensor1D::zeros().traced());
    let gradients = y.square().mean().backward();

    let mut opt: Sgd<Model> = Default::default();
    opt.update(&mut model, gradients).expect("");
}
```

Which fails with the message:
```
thread 'main' panicked at ': UnusedParamsError { param_locations: ["0.0.0.weight", "0.0.0.bias", "0.0.1.weight", "0.0.1.bias", "0.0.2.weight", "0.0.2.bias"] }', examples\bad.rs:14:39
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: process didn't exit successfully: `target\debug\examples\bad.exe` (exit code: 101)
```

Definitely isn't super easy to tell what parameters its talking about though...

Reply:
Yes, but it is still better than having the same names repeated. This is one area where a language like Python with it's reflection shines, it can just lookup the name of the module class and prepend it.

But I say this looks good for now.